// ng2-konva/esm5/lib/konva.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { StageComponent } from './components/stage.component';
import { CoreShapeComponent } from './components/core-shape.component';
if (typeof window !== 'undefined' && !window.Konva) {
    require('konva');
}
export { StageComponent } from './components/stage.component';
export { CoreShapeComponent } from './components/core-shape.component';
var KonvaModule = /** @class */ (function () {
    function KonvaModule() {
    }
    KonvaModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule
                    ],
                    declarations: [
                        StageComponent,
                        CoreShapeComponent,
                    ],
                    exports: [
                        StageComponent,
                        CoreShapeComponent,
                    ]
                },] },
    ];
    return KonvaModule;
}());
export { KonvaModule };
