// ng2-konva/esm2015/lib/components/stage.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ElementRef, ContentChildren, QueryList } from '@angular/core';
import { CoreShapeComponent as CoreShape } from './core-shape.component';
import { Observable } from 'rxjs';
import { updatePicture, createListener, applyNodeProps } from '../utils/index';
export class StageComponent {
    /**
     * @param {?} elementRef
     */
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.shapes = new QueryList();
        this.click = new EventEmitter();
        this.dblclick = new EventEmitter();
        this.mouseover = new EventEmitter();
        this.mouseout = new EventEmitter();
        this.mousemove = new EventEmitter();
        this.tap = new EventEmitter();
        this.dbltap = new EventEmitter();
        this.touchstart = new EventEmitter();
        this.scaleXChange = new EventEmitter();
        this.fillChange = new EventEmitter();
        this.dragstart = new EventEmitter();
        this.dragmove = new EventEmitter();
        this.dragend = new EventEmitter();
        this.cacheProps = {};
    }
    /**
     * @return {?}
     */
    getStage() {
        return this._stage || {};
    }
    /**
     * @return {?}
     */
    getConfig() {
        return this._config || {};
    }
    /**
     * @param {?} config
     * @return {?}
     */
    uploadKonva(config) {
        const /** @type {?} */ props = Object.assign({}, config, createListener(this));
        applyNodeProps(this, props, this.cacheProps);
        this.cacheProps = props;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        const /** @type {?} */ nodeContainer = this.elementRef.nativeElement;
        this.config.subscribe(config => {
            this._config = config;
            if (!this._stage) {
                this._stage = new Konva.Stage({
                    width: config.width,
                    height: config.height,
                    container: nodeContainer
                });
                this.uploadKonva(config);
            }
            else {
                this.uploadKonva(config);
            }
        });
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this.shapes.forEach((item) => {
            this._stage.add(item.getStage());
            updatePicture(this._stage);
        });
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this._stage.destroy();
    }
}
StageComponent.decorators = [
    { type: Component, args: [{
                selector: 'ko-stage',
                template: `<div><ng-content></ng-content>{{config}}</div>`
            },] },
];
/** @nocollapse */
StageComponent.ctorParameters = () => [
    { type: ElementRef }
];
StageComponent.propDecorators = {
    shapes: [{ type: ContentChildren, args: [CoreShape,] }],
    config: [{ type: Input }],
    click: [{ type: Output }],
    dblclick: [{ type: Output }],
    mouseover: [{ type: Output }],
    mouseout: [{ type: Output }],
    mousemove: [{ type: Output }],
    tap: [{ type: Output }],
    dbltap: [{ type: Output }],
    touchstart: [{ type: Output }],
    scaleXChange: [{ type: Output }],
    fillChange: [{ type: Output }],
    dragstart: [{ type: Output }],
    dragmove: [{ type: Output }],
    dragend: [{ type: Output }]
};
function StageComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    StageComponent.prototype.shapes;
    /** @type {?} */
    StageComponent.prototype.config;
    /** @type {?} */
    StageComponent.prototype.click;
    /** @type {?} */
    StageComponent.prototype.dblclick;
    /** @type {?} */
    StageComponent.prototype.mouseover;
    /** @type {?} */
    StageComponent.prototype.mouseout;
    /** @type {?} */
    StageComponent.prototype.mousemove;
    /** @type {?} */
    StageComponent.prototype.tap;
    /** @type {?} */
    StageComponent.prototype.dbltap;
    /** @type {?} */
    StageComponent.prototype.touchstart;
    /** @type {?} */
    StageComponent.prototype.scaleXChange;
    /** @type {?} */
    StageComponent.prototype.fillChange;
    /** @type {?} */
    StageComponent.prototype.dragstart;
    /** @type {?} */
    StageComponent.prototype.dragmove;
    /** @type {?} */
    StageComponent.prototype.dragend;
    /** @type {?} */
    StageComponent.prototype._stage;
    /** @type {?} */
    StageComponent.prototype._config;
    /** @type {?} */
    StageComponent.prototype.cacheProps;
    /** @type {?} */
    StageComponent.prototype.elementRef;
}
