// ng2-konva/esm5/lib/components/stage.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, Output, EventEmitter, ElementRef, ContentChildren, QueryList } from '@angular/core';
import { CoreShapeComponent as CoreShape } from './core-shape.component';
import { Observable } from 'rxjs';
import { updatePicture, createListener, applyNodeProps } from '../utils/index';
var StageComponent = /** @class */ (function () {
    function StageComponent(elementRef) {
        this.elementRef = elementRef;
        this.shapes = new QueryList();
        this.click = new EventEmitter();
        this.dblclick = new EventEmitter();
        this.mouseover = new EventEmitter();
        this.mouseout = new EventEmitter();
        this.mousemove = new EventEmitter();
        this.tap = new EventEmitter();
        this.dbltap = new EventEmitter();
        this.touchstart = new EventEmitter();
        this.scaleXChange = new EventEmitter();
        this.fillChange = new EventEmitter();
        this.dragstart = new EventEmitter();
        this.dragmove = new EventEmitter();
        this.dragend = new EventEmitter();
        this.cacheProps = {};
    }
    /**
     * @return {?}
     */
    StageComponent.prototype.getStage = /**
     * @return {?}
     */
    function () {
        return this._stage || {};
    };
    /**
     * @return {?}
     */
    StageComponent.prototype.getConfig = /**
     * @return {?}
     */
    function () {
        return this._config || {};
    };
    /**
     * @param {?} config
     * @return {?}
     */
    StageComponent.prototype.uploadKonva = /**
     * @param {?} config
     * @return {?}
     */
    function (config) {
        var /** @type {?} */ props = tslib_1.__assign({}, config, createListener(this));
        applyNodeProps(this, props, this.cacheProps);
        this.cacheProps = props;
    };
    /**
     * @return {?}
     */
    StageComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        var /** @type {?} */ nodeContainer = this.elementRef.nativeElement;
        this.config.subscribe(function (config) {
            _this._config = config;
            if (!_this._stage) {
                _this._stage = new Konva.Stage({
                    width: config.width,
                    height: config.height,
                    container: nodeContainer
                });
                _this.uploadKonva(config);
            }
            else {
                _this.uploadKonva(config);
            }
        });
    };
    /**
     * @return {?}
     */
    StageComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.shapes.forEach(function (item) {
            _this._stage.add(item.getStage());
            updatePicture(_this._stage);
        });
    };
    /**
     * @return {?}
     */
    StageComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this._stage.destroy();
    };
    StageComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ko-stage',
                    template: "<div><ng-content></ng-content>{{config}}</div>"
                },] },
    ];
    /** @nocollapse */
    StageComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    StageComponent.propDecorators = {
        shapes: [{ type: ContentChildren, args: [CoreShape,] }],
        config: [{ type: Input }],
        click: [{ type: Output }],
        dblclick: [{ type: Output }],
        mouseover: [{ type: Output }],
        mouseout: [{ type: Output }],
        mousemove: [{ type: Output }],
        tap: [{ type: Output }],
        dbltap: [{ type: Output }],
        touchstart: [{ type: Output }],
        scaleXChange: [{ type: Output }],
        fillChange: [{ type: Output }],
        dragstart: [{ type: Output }],
        dragmove: [{ type: Output }],
        dragend: [{ type: Output }]
    };
    return StageComponent;
}());
export { StageComponent };
function StageComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    StageComponent.prototype.shapes;
    /** @type {?} */
    StageComponent.prototype.config;
    /** @type {?} */
    StageComponent.prototype.click;
    /** @type {?} */
    StageComponent.prototype.dblclick;
    /** @type {?} */
    StageComponent.prototype.mouseover;
    /** @type {?} */
    StageComponent.prototype.mouseout;
    /** @type {?} */
    StageComponent.prototype.mousemove;
    /** @type {?} */
    StageComponent.prototype.tap;
    /** @type {?} */
    StageComponent.prototype.dbltap;
    /** @type {?} */
    StageComponent.prototype.touchstart;
    /** @type {?} */
    StageComponent.prototype.scaleXChange;
    /** @type {?} */
    StageComponent.prototype.fillChange;
    /** @type {?} */
    StageComponent.prototype.dragstart;
    /** @type {?} */
    StageComponent.prototype.dragmove;
    /** @type {?} */
    StageComponent.prototype.dragend;
    /** @type {?} */
    StageComponent.prototype._stage;
    /** @type {?} */
    StageComponent.prototype._config;
    /** @type {?} */
    StageComponent.prototype.cacheProps;
    /** @type {?} */
    StageComponent.prototype.elementRef;
}
